src/jvmMain/kotlin/environments/SeEnvironment.kt

```diff
@@ -0,0 +1,133 @@
+}
+
+val WorldEntity.integrity: Int
+    get() {
+        return properties["integrity"] as Int
+    }
+
+val WorldEntity.maxIntegrity: Int
+    get() {
+        return properties["maxIntegrity"] as Int
+    }
+
+val WorldEntity.buildIntegrity: Int
+    get() {
+        return properties["buildIntegrity"] as Int
+    }
+
+val WorldEntity.blockType: String
+    get() {
+        return properties["blockType"] as String
+    }
+
+fun SeGrid.toWorldEntity(): WorldEntity {
+    return WorldEntity(this.id, "grid", false).also { we ->
+        we.position = position.toIv4xrVec3()
+        we.elements = blocks.map { it.toWorldEntity() }.associateBy { it.id }
+    }
+}
+
+fun SeBlock.toWorldEntity(): WorldEntity {
+    return WorldEntity(id, "block", false).also { we ->
+        we.position = position.toIv4xrVec3()
+        we.properties["blockType"] = blockType
+        we.properties["integrity"] = integrity
+        we.properties["maxIntegrity"] = maxIntegrity
+        we.properties["buildIntegrity"] = buildIntegrity
+        size?.let {
+            we.extent = it.toIv4xrVec3()
+        }
+    }
+}
+
+fun SeObservation.toWorldModel(): WorldModel {
+    return WorldModel().also { worldModel ->
+        worldModel.agentId = agentID
+        worldModel.position = position.toIv4xrVec3()
+        worldModel.velocity = velocity.toIv4xrVec3()
+        worldModel.elements = grids?.map { it.toWorldEntity() }?.associateBy { it.id } ?: emptyMap()
+    }
+}
+
+val W3DAgentState.seEnv: SeEnvironment
+    get() = env() as SeEnvironment
+
+fun W3DAgentState.setOrUpdate(worldModel: WorldModel) {
+    if (wom == null) {
+        wom = worldModel
+    } else {
+        wom.mergeNewObservation(worldModel)
+    }
+}
+
+fun W3DAgentState.observe() {
+    setOrUpdate(seEnv.observe())
+}
+
+class SeEnvironment(
+    val worldId: String,
+    val controller: ContextControllerWrapper,
+    val worldController: WorldController,
+    val context: SpaceEngineersTestContext
+) : W3DEnvironment() {
+    val SCENARIO_DIR = "src/jvmTest/resources/game-saves/"
+
+
+    override fun loadWorld() {
+        val scenario = File("$SCENARIO_DIR$worldId").absolutePath
+        worldController.load(scenario)
+    }
+
+    override fun observe(agentId: String): WorldModel {
+        return controller.observe().toWorldModel()
+    }
+
+    fun observe(): WorldModel {
+        return controller.observe().toWorldModel()
+    }
+
+    fun observeForNewBlocks(): WorldModel {
+        return controller.observeNewBlocks().toWorldModel()
+    }
+
+    fun equipAndPlace(toolbarLocation: ToolbarLocation): WorldModel {
+        return controller.equipAndPlace(toolbarLocation).toWorldModel()
+    }
+
+    fun equip(toolbarLocation: ToolbarLocation): WorldModel {
+        return controller.equip(toolbarLocation).toWorldModel()
+    }
+
+    fun startUsingTool(): WorldModel {
+        controller.startUsingTool().toWorldModel()
+        return observe()
+    }
```

Comment:
I don't understand this much. Why there is the call .toWorldModel(), if you just throw it away?

Why there is separate observe() call? It's because you want an observation after the action is applied?

That means we would probably change the plugin API not to return any observation by default.

Reply:
* Re toWorldModel converts SeObservation to WorldModel object so I am just following our current API. But yeah it's unnecessary since I am returning observation from observe.
* Re observe() - yes, maybe unnecessary in this case.
* Definitely, or return something useful. This method can fail if ran too soon after equip (that's why the sleep) so it would be nice to return something that will show success/failure or throw exception if it fails (but our current transport layer doesn't really support/handle errors well).

Overall I think it's mostly confusing, because our API is bad and sensible return value would fix it.

Reply:
So please remove the unnecessary toWorldModel call.

And comment on the other thing with the return value somewhere.

I see you are reluctant to add any comments, but I think the world will never be so ideal that everything is intuitive and without any need for comments.